core/src/main/java/io/grpc/internal/AbstractManagedChannelImplBuilder.java

```diff
@@ -99,4 +162,50 @@ public ManagedChannelImpl build() {
    */
   @Internal
   protected abstract ClientTransportFactory buildTransportFactory();
+
+  private static class DirectAddressNameResolver implements NameResolver {
+    final SocketAddress address;
+
+    DirectAddressNameResolver(SocketAddress address) {
+      this.address = address;
+    }
+
+    @Override
+    public Handle resolve(String scheme, @Nullable String nsAuthority, String endpoint) {
+      return new Handle() {
+        @Override
+        public String getServiceAuthority() {
+          if (address instanceof InetSocketAddress) {
```

Comment:
This _really_ needs to come from the transport or we need to wire in authority overriding. The authority returned here should be the same authority used for TLS, and that authority may be different from `endpoint` for use in testing.